src/pushed_authorization.rs

```diff
@@ -0,0 +1,293 @@
+            user_hint,
+            issuer_state,
+        }
+    }
+
+    pub async fn async_request<C, F, RE>(
+        self,
+        mut http_client: C,
+        client_assertion_type: Option<String>,
+        client_assertion: Option<String>,
+    ) -> Result<(url::Url, CsrfToken), RequestError<RE>>
+    where
+        C: FnMut(HttpRequest) -> F,
+        F: Future<Output = Result<HttpResponse, RE>>,
+        RE: std::error::Error + 'static,
+    {
+        let mut auth_url = self.auth_url.url().clone();
+
+        let (http_request, req_body, token) =
+            self.prepare_request(client_assertion, client_assertion_type)?;
+
+        let http_response = http_client(http_request)
+            .await
+            .map_err(RequestError::Request)?;
+
+        if http_response.status_code != StatusCode::OK {
+            return Err(RequestError::Response(
+                http_response.status_code,
+                http_response.body,
+                "unexpected HTTP status code".to_string(),
+            ));
+        }
+
+        let parsed_response: PushedAuthorizationResponse = match http_response
+            .headers
+            .get(CONTENT_TYPE)
+            .map(ToOwned::to_owned)
+            .unwrap_or_else(|| HeaderValue::from_static(MIME_TYPE_JSON))
+        {
+            ref content_type if content_type_has_essence(content_type, MIME_TYPE_JSON) => {
+                serde_path_to_error::deserialize(&mut serde_json::Deserializer::from_slice(
+                    &http_response.body,
+                ))
+                .map_err(RequestError::Parse)
+            }
+            ref content_type => Err(RequestError::Response(
+                http_response.status_code,
+                http_response.body,
+                format!("unexpected response Content-Type: `{:?}`", content_type),
+            )),
+        }?;
+
+        auth_url
+            .query_pairs_mut()
+            .append_pair("request_uri", parsed_response.request_uri.get());
+
+        auth_url
+            .query_pairs_mut()
+            .append_pair("client_id", &req_body.client_id.to_string());
+
+        Ok((auth_url, token))
+    }
+
+    pub fn prepare_request<RE>(
+        self,
+        client_assertion: Option<String>,
+        client_assertion_type: Option<String>,
+    ) -> Result<(HttpRequest, ParAuthParams, CsrfToken), RequestError<RE>>
+    where
+        RE: std::error::Error + 'static,
+    {
+        let (url, token) = self.inner.url();
+
+        let body = serde_urlencoded::from_str::<ParAuthParams>(url.clone().as_str())
+            .map_err(|_| RequestError::Other("failed parsing url".to_string()))?
+            .set_client_assertion_type(client_assertion_type.clone())
+            .set_client_assertion(client_assertion.clone())
+            .set_authorization_details(Some(
+                serde_json::to_string::<Vec<AuthorizationDetail<AD>>>(&self.authorization_details)
+                    .unwrap_or("[]".to_string()),
```

Comment:
I think we should return an error rather than silently failing
```suggestion
                serde_json::to_string::<Vec<AuthorizationDetail<AD>>>(&self.authorization_details)
                    .map_err(|e| RequestError::Other(format!("unable to serialize authorization_details: {}", e)))?,
```